### PR TITLE
Pod decorator extension point

### PR DESCRIPTION
**Pull request by Vlatombe**
Trying a replacement of some proprietary `PodTemplateFilter` implementations

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodDecorator.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import hudson.ExtensionPoint;
+import io.fabric8.kubernetes.api.model.PodBuilder;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Allows to alter a pod definition after it has been built from the yaml and DSL/GUI configuration.
+ */
+public abstract class PodDecorator implements ExtensionPoint {
+    public static ExtensionList<PodDecorator> all() {
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodDecorator.java:13]: Just inline.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodDecorator.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import io.fabric8.kubernetes.api.model.PodBuilder;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Allows to alter a pod definition after it has been built from the yaml and DSL/GUI configuration.
+ */
+public abstract class PodDecorator implements ExtensionPoint {
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodDecorator.java:12]: Could be an `interface`.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 3]
```diff
@@ -233,7 +233,7 @@ public Pod build() {
         }
 
         // merge with the yaml fragments
-        Pod pod = combine(template.getYamlsPod(), builder.endSpec().build());
+        Pod pod = PodDecorator.decorateAll(new PodBuilder(combine(template.getYamlsPod(), builder.endSpec().build()))).build();
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:236]: Does this cost anything if the list is empty?

Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java]: It's pretty cheap but I refactored to avoid it.

**Review by jglick — APPROVED** [review]
Looks nice. Now just needs a functional test and a sample implementation.

**Comment by Vlatombe** [comment]
I'm hesitating to add `PodTemplate` to the extension point. It could be useful in particular to implement [JENKINS-59909](https://issues.jenkins.io/browse/JENKINS-59909)

**Review by jglick — APPROVED** [review]
Looks reasonable.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java [hunk 4]
```diff
@@ -301,6 +301,7 @@ private void validatePod(Pod pod, boolean directConnection) {
 
     private void validatePod(Pod pod, boolean fromYaml, boolean directConnection) {
         assertThat(pod.getMetadata().getLabels(), hasEntry("some-label", "some-label-value"));
+        assertEquals("Never", pod.getSpec().getRestartPolicy());
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java:304]: https://github.com/jenkinsci/kubernetes-plugin/pull/906/checks?check_run_id=1473530018

Comment (Vlatombe) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilderTest.java:304]: yeah, looking into it.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 5]
```diff
@@ -166,7 +168,7 @@ public Pod build() {
             }
         }
 
-        volumes.put(WORKSPACE_VOLUME_NAME, template.getWorkspaceVolume().buildVolume(WORKSPACE_VOLUME_NAME, slave != null ? slave.getPodName() : null));
+        volumes.put(WORKSPACE_VOLUME_NAME, template.getWorkspaceVolume().buildVolume(WORKSPACE_VOLUME_NAME, agent != null ? agent.getPodName() : null));
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:176]:
```suggestion
        volumes.put(WORKSPACE_VOLUME_NAME, template.getWorkspaceVolume().buildVolume(WORKSPACE_VOLUME_NAME, agent.getPodName()));
```

if it is really `@Nonnull`?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java [hunk 6]
```diff
@@ -176,13 +178,13 @@ public Pod build() {
         }
 
         MetadataNested<PodBuilder> metadataBuilder = new PodBuilder().withNewMetadata();
-        if (slave != null) {
-            metadataBuilder.withName(slave.getPodName());
+        if (agent != null) {
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:186]: Ditto here, etc.?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java (hunk 5 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java:176]: The old constructor is used in other places so I have to keep it `@CheckForNull`

**Pull request merged**